apps/near_app/near_pub_key.c

```diff
@@ -0,0 +1,428 @@
+  if (which_request != query->get_public_key.which_request) {
+    near_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                    ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request(const near_get_public_key_intiate_request_t *req,
+                             const pb_size_t which_request) {
+  bool status = true;
+  const pb_size_t count = req->derivation_paths_count;
+
+  if (0 == count) {
+    // request does not have any derivation paths, invalid request
+    near_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                    ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+
+  if (NEAR_QUERY_GET_PUBLIC_KEY_TAG == which_request && 1 < count) {
+    // `GET_PUBLIC_KEY` request contains more than one derivation path which is
+    // not expected
+    near_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                    ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+
+  const near_get_public_key_derivation_path_t *path = NULL;
+
+  for (pb_size_t index = 0; index < count; index++) {
+    path = &req->derivation_paths[index];
+    if (!near_derivation_path_guard(path->path, path->path_count)) {
+      near_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                      ERROR_DATA_FLOW_INVALID_DATA);
+      status = false;
+      break;
+    }
+  }
+
+  return status;
+}
+
+static bool get_public_key(const uint8_t *seed,
+                           const uint32_t *path,
+                           uint32_t path_length,
+                           uint8_t *public_key) {
+  HDNode node = {0};
+
+  if (!derive_hdnode_from_path(path, path_length, ED25519_NAME, seed, &node)) {
+    memzero(&node, sizeof(HDNode));
+    return false;
+  }
+
+  if (NULL != public_key) {
+    ecdsa_uncompress_pubkey(
+        get_curve_by_name(ED25519_NAME)->params, node.public_key, public_key);
+  }
+
+  memzero(&node, sizeof(HDNode));
+  return true;
+}
+
+static void fill_public_keys(const near_get_public_key_derivation_path_t *paths,
+                             const uint8_t *seed,
+                             uint8_t public_key_list[][NEAR_PUB_KEY_SIZE],
+                             pb_size_t count) {
+  for (pb_size_t index = 0; index < count; index++) {
+    const near_get_public_key_derivation_path_t *path = &paths[index];
+    if (!get_public_key(
+            seed, path->path, path->path_count, public_key_list[index])) {
+      // send unknown error; unknown failure reason
+      near_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static bool send_public_keys(near_query_t *query,
+                             const uint8_t pubkey_list[][NEAR_PUB_KEY_SIZE],
+                             const pb_size_t count,
+                             const pb_size_t which_request) {
+  near_result_t response = init_near_result(NEAR_RESULT_GET_PUBLIC_KEY_TAG);
```

Comment:
Verify if the logic acc. to proto files is the same for get public keys and receive operation.

Reply:
Thanks for this catch